packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/agent_manager.ts

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { Agent as HttpAgent } from 'http';
+import { Agent as HttpsAgent } from 'https';
+import { ConnectionOptions, HttpAgentOptions, UndiciAgentOptions } from '@elastic/elasticsearch';
+import { Logger } from '@kbn/logging';
+
+const HTTP = 'http:';
+const HTTPS = 'https:';
+
+export type Protocol = typeof HTTP | typeof HTTPS;
+export type NetworkAgent = HttpAgent | HttpsAgent;
+export type AgentFactory = (connectionOpts: ConnectionOptions) => NetworkAgent;
+
+/**
+ * Allows obtaining Agent factories, which can then be fed into elasticsearch-js's Client class.
+ * Ideally, we should obtain one Agent factory for each ES Client class.
+ * This allows using the same Agent across all the Pools and Connections of the Client (one per ES node).
+ *
+ * Agent instances are stored internally to allow collecting metrics (nbr of active/idle connections to ES).
+ *
+ * Using the same Agent factory across multiple ES Client instances is strongly discouraged, cause ES Client
+ * exposes methods that can modify the underlying pools, effectively impacting the connections of other Clients.
+ * @internal
+ **/
+export class AgentManager {
+  /**
+   * Store Agent instances by type and protocol, e.g.:
+   *
+   * data: {
+   *   'http:': [agentInstance1, agentInstance2],
+   *   'https:': [agentInstance3, agentInstance4]
+   * },
+   * monitoring: {
+   *   'http:': [],
+   *   'https:': [agentInstance5]
+   * }
+   */
+  private agentStore: Record<string, Record<Protocol, Array<NetworkAgent | undefined>>>;
+  private logger: Logger;
+
+  constructor(
+    logger: Logger,
+    private defaultConfig: HttpAgentOptions = {
+      keepAlive: true,
+      keepAliveMsecs: 50000,
+      maxSockets: 256,
+      maxFreeSockets: 256,
+      scheduling: 'lifo',
+    }
+  ) {
+    this.logger = logger;
+    this.agentStore = {};
+  }
+
+  public getAgentFactory(
+    type: string,
+    agentOptions?: HttpAgentOptions | UndiciAgentOptions | AgentFactory | false
+  ): AgentFactory {
+    if (isAgentFactory(agentOptions)) {
+      // use the user-provided factory directly
+      return agentOptions;
+    }
```

Comment:
From the client's type def:

``export declare type agentFn = (opts: ConnectionOptions) => any;``

Nice... Helps a lot...

Do you know what an `agentFn` is actually supposed to return? Can we wrap the return or fn with a HOF somehow to wire our logic / would it make sense?

Reply:
`agentFn` is supposed to return `Agent` instances.
I created the equivalent `AgentFactory`, and the `NetworkAgent` types that are a bit more specific.

Reply:
IMO it doesn't make sense to create an agent factory and pass in a custom factory, then we should just use the custom factory. I suspect it might be because you wanted to match the Elasticsearch-js config options type which includes being able to specify a factory? But plugins cannot create Elasticsearch-js clients with custom factories because we only expose a limited set of options https://github.com/elastic/kibana/blob/main/packages/core/elasticsearch/core-elasticsearch-server/src/client/client_config.ts#L16-L33

To be defensive, could we just throw if agent options is a function?

Reply:
Yeah, ATM the `agent?: HttpAgentOptions | UndiciAgentOptions | agentFn | false;` can only be `HttpAgentOptions` according to how we build it in [parseClientOptions](https://github.com/elastic/kibana/blob/main/packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/client_config.ts#L37,L40).

How about we make the internal API typings more specific and adjusted to reality?

Reply:
If you can restrict the types to exclude `agentFn` it would work, but given `parseClientOptions` returns a plain `ClientOptions`, I don't think you would be able without some type shenanigans or force-casting, so throwing seems fine too (at least preferable than pass-through-ing as the impl currently does).